# Assignment #2 - Classification

<font color="red"> <b> Due: April 10  EOD</b> </font>

<font color="blue"> type your name here </font>

# I. Introduction

Describe the objective of this assignment. You can briefly state how you accompilsh it.

# II. Data

You may choose one of the following three datasets to work on. Introduce your data and visualize them. Describe your observations about the data. 
1. https://archive.ics.uci.edu/ml/datasets/covertype
2. https://archive.ics.uci.edu/ml/datasets/bank+marketing
3. https://archive.ics.uci.edu/ml/datasets/Activity+recognition+using+wearable+physiological+measurements (data download link - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6960825/bin/sensors-19-05524-s001.zip)


# III. Method

Summarize the Least Mean Square Classifier, Fisher Linear Discriminant, Perceptron,logistic regression, and Neural Network.
Do not forget explain your implementation. 

The explanation of your codes should not be the comments in a code cell. 
Each implementation will be followed by a separate cell that should include
 - review of the 4 classification models 
 - your implementation and description


### Method

In [2]:
# importing important libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_curve, roc_auc_score
from sklearn.preprocessing import StandardScaler,LabelEncoder
import matplotlib.pyplot as plt
from prml.linear import (
    BayesianLogisticRegression,
    LeastSquaresClassifier,
    FishersLinearDiscriminant,
    LogisticRegression,
    Perceptron,
    SoftmaxRegression
)

np.random.seed(1234)

##your method implementation goes here

In [3]:
# reading dataset
df = pd.read_csv('bank-additional.csv')
df.head(10)


age           job   marital            education  default  housing  \
0   30   blue-collar   married             basic.9y       no      yes   
1   39      services    single          high.school       no       no   
2   25      services   married          high.school       no      yes   
3   38      services   married             basic.9y       no  unknown   
4   47        admin.   married    university.degree       no      yes   
5   32      services    single    university.degree       no       no   
6   32        admin.    single    university.degree       no      yes   
7   41  entrepreneur   married    university.degree  unknown      yes   
8   31      services  divorced  professional.course       no       no   
9   35   blue-collar   married             basic.9y  unknown       no   

      loan    contact month day_of_week  ...  campaign  pdays  previous  \
0       no   cellular   may         fri  ...         2    999         0   
1       no  telephone   may         fri  ...         4    999         0   
2       no  telephone   jun         wed  ...         1    999         0   
3  unknown  telephone   jun         fri  ...         3    999         0   
4       no   cellular   nov         mon  ...         1    999         0   
5       no   cellular   sep         thu  ...         3    999         2   
6       no   cellular   sep         mon  ...         4    999         0   
7       no   cellular   nov         mon  ...         2    999         0   
8       no   cellular   nov         tue  ...         1    999         1   
9       no  telephone   may         thu  ...         1    999         0   

      poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0  nonexistent         -1.8          92.893          -46.2      1.313   
1  nonexistent          1.1          93.994          -36.4      4.855   
2  nonexistent          1.4          94.465          -41.8      4.962   
3  nonexistent          1.4          94.465          -41.8      4.959   
4  nonexistent         -0.1          93.200          -42.0      4.191   
5      failure         -1.1          94.199          -37.5      0.884   
6  nonexistent         -1.1          94.199          -37.5      0.879   
7  nonexistent         -0.1          93.200          -42.0      4.191   
8      failure         -0.1          93.200          -42.0      4.153   
9  nonexistent          1.1          93.994          -36.4      4.855   

   nr.employed   y  
0       5099.1  no  
1       5191.0  no  
2       5228.1  no  
3       5228.1  no  
4       5195.8  no  
5       4963.6  no  
6       4963.6  no  
7       5195.8  no  
8       5195.8  no  
9       5191.0  no  

[10 rows x 21 columns]

### Some variables have binary values hence using One Hot Encoding

In [4]:

df = pd.get_dummies(df, columns=['marital', 'default','loan','housing', 'contact','poutcome'])
df.head()

age          job          education month day_of_week  duration  campaign  \
0   30  blue-collar           basic.9y   may         fri       487         2   
1   39     services        high.school   may         fri       346         4   
2   25     services        high.school   jun         wed       227         1   
3   38     services           basic.9y   jun         fri        17         3   
4   47       admin.  university.degree   nov         mon        58         1   

   pdays  previous  emp.var.rate  ...  loan_unknown  loan_yes  housing_no  \
0    999         0          -1.8  ...             0         0           0   
1    999         0           1.1  ...             0         0           1   
2    999         0           1.4  ...             0         0           0   
3    999         0           1.4  ...             1         0           0   
4    999         0          -0.1  ...             0         0           0   

   housing_unknown housing_yes  contact_cellular  contact_telephone  \
0                0           1                 1                  0   
1                0           0                 0                  1   
2                0           1                 0                  1   
3                1           0                 0                  1   
4                0           1                 1                  0   

   poutcome_failure  poutcome_nonexistent  poutcome_success  
0                 0                     1                 0  
1                 0                     1                 0  
2                 0                     1                 0  
3                 0                     1                 0  
4                 0                     1                 0  

[5 rows x 33 columns]

### Some variables have ordinal values hence using Label encoding

In [5]:

labelEncoder = LabelEncoder() 

df['education'] = labelEncoder.fit_transform(df['education'])
df['job'] = labelEncoder.fit_transform(df['job'])
df['month'] = labelEncoder.fit_transform(df['month'])
df['y'] = labelEncoder.fit_transform(df['y'])
df['day_of_week'] = labelEncoder.fit_transform(df['day_of_week'])
df['pdays'] = labelEncoder.fit_transform(df['pdays'])




In [6]:
df.head(20)

age  job  education  month  day_of_week  duration  campaign  pdays  \
0    30    1          2      6            0       487         2     20   
1    39    7          3      6            0       346         4     20   
2    25    7          3      4            4       227         1     20   
3    38    7          2      4            0        17         3     20   
4    47    0          6      7            1        58         1     20   
5    32    7          6      9            2       128         3     20   
6    32    0          6      9            1       290         4     20   
7    41    2          6      7            1        44         2     20   
8    31    7          5      7            3        68         1     20   
9    35    1          2      6            2       170         1     20   
10   25    7          1      3            2       301         1     20   
11   36    6          0      3            2       148         1     20   
12   36    0          3      6            4        97         2     20   
13   47    1          0      4            2       211         2     20   
14   29    0          3      6            0       553         2     20   
15   27    7          6      3            4       698         2     20   
16   44    0          6      3            4       191         6     20   
17   46    0          6      3            1        59         4     20   
18   45    2          6      1            1        38         2     20   
19   50    1          0      3            3       849         1     20   

    previous  emp.var.rate  ...  loan_unknown  loan_yes  housing_no  \
0          0          -1.8  ...             0         0           0   
1          0           1.1  ...             0         0           1   
2          0           1.4  ...             0         0           0   
3          0           1.4  ...             1         0           0   
4          0          -0.1  ...             0         0           0   
5          2          -1.1  ...             0         0           1   
6          0          -1.1  ...             0         0           0   
7          0          -0.1  ...             0         0           0   
8          1          -0.1  ...             0         0           1   
9          0           1.1  ...             0         0           1   
10         0           1.4  ...             0         0           0   
11         0           1.4  ...             0         0           1   
12         0           1.1  ...             0         0           1   
13         0           1.4  ...             0         0           0   
14         0          -1.8  ...             0         0           1   
15         0           1.4  ...             0         0           1   
16         0           1.4  ...             0         0           1   
17         0           1.4  ...             0         0           0   
18         0           1.4  ...             0         1           0   
19         0           1.4  ...             0         1           1   

    housing_unknown  housing_yes  contact_cellular  contact_telephone  \
0                 0            1                 1                  0   
1                 0            0                 0                  1   
2                 0            1                 0                  1   
3                 1            0                 0                  1   
4                 0            1                 1                  0   
5                 0            0                 1                  0   
6                 0            1                 1                  0   
7                 0            1                 1                  0   
8                 0            0                 1                  0   
9                 0            0                 0                  1   
10                0            1                 1                  0   
11                0            0                 1                  0   
12      

In [7]:
df.head()

age  job  education  month  day_of_week  duration  campaign  pdays  \
0   30    1          2      6            0       487         2     20   
1   39    7          3      6            0       346         4     20   
2   25    7          3      4            4       227         1     20   
3   38    7          2      4            0        17         3     20   
4   47    0          6      7            1        58         1     20   

   previous  emp.var.rate  ...  loan_unknown  loan_yes  housing_no  \
0         0          -1.8  ...             0         0           0   
1         0           1.1  ...             0         0           1   
2         0           1.4  ...             0         0           0   
3         0           1.4  ...             1         0           0   
4         0          -0.1  ...             0         0           0   

   housing_unknown  housing_yes  contact_cellular  contact_telephone  \
0                0            1                 1                  0   
1                0            0                 0                  1   
2                0            1                 0                  1   
3                1            0                 0                  1   
4                0            1                 1                  0   

   poutcome_failure  poutcome_nonexistent  poutcome_success  
0                 0                     1                 0  
1                 0                     1                 0  
2                 0                     1                 0  
3                 0                     1                 0  
4                 0                     1                 0  

[5 rows x 33 columns]

In [8]:
X = df.drop(columns=['y'])
y = df['y']


### Spliting the dataset and appplying scaling 

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((2883, 32), (1236, 32), (2883,), (1236,))

In [10]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

### 1. Least Square Classifier

In [11]:
model = LeastSquaresClassifier()
model.fit(X_train, y_train)
y_pred = model.classify(X_test)


In [12]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print("Model Accuracy %:", accuracy_score(y_test, y_pred)*100)
print()
print("Classification Report :",classification_report(y_test,y_pred))
print()
print("Confusion Matrix :",confusion_matrix(y_test,y_pred))
 

Model Accuracy %: 78.39805825242718

Classification Report :               precision    recall  f1-score   support

           0       0.99      0.76      0.86      1105
           1       0.32      0.96      0.49       131

    accuracy                           0.78      1236
   macro avg       0.66      0.86      0.67      1236
weighted avg       0.92      0.78      0.82      1236


Confusion Matrix : [[843 262]
 [  5 126]]


### 2 .Logistic Regression

In [13]:
model1 = LogisticRegression()
model1.fit(X_train, y_train)
y_pred1 = model1.classify(X_test)


In [14]:
print("Model Accuracy %:", accuracy_score(y_test, y_pred1)*100)
print()
print("Classification Report :",classification_report(y_test,y_pred1))
print()
print("Confusion Matrix :",confusion_matrix(y_test,y_pred1))
 

Model Accuracy %: 89.40129449838187

Classification Report :               precision    recall  f1-score   support

           0       0.89      1.00      0.94      1105
           1       0.00      0.00      0.00       131

    accuracy                           0.89      1236
   macro avg       0.45      0.50      0.47      1236
weighted avg       0.80      0.89      0.84      1236


Confusion Matrix : [[1105    0]
 [ 131    0]]


C:\Users\dell\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## 3.Softmax Regression

In [15]:
model2 = SoftmaxRegression()
model2.fit(X_train, y_train)
y_pred2 = model2.classify(X_test)


In [16]:
print("Model Accuracy %:", accuracy_score(y_test, y_pred2)*100)
print()
print("Classification Report :",classification_report(y_test,y_pred2))
print()
print("Confusion Matrix :",confusion_matrix(y_test,y_pred2))
 

Model Accuracy %: 62.45954692556634

Classification Report :               precision    recall  f1-score   support

           0       0.95      0.61      0.74      1105
           1       0.18      0.73      0.29       131

    accuracy                           0.62      1236
   macro avg       0.57      0.67      0.52      1236
weighted avg       0.87      0.62      0.70      1236


Confusion Matrix : [[677 428]
 [ 36  95]]


## 4. BayesianLogisticRegression

In [17]:
model3 = BayesianLogisticRegression()
model3.fit(X_train, y_train)
y_pred3 = model3.classify(X_test)

In [20]:
print("Model Accuracy %:", accuracy_score(y_test, y_pred3)*100)
print()
print("Classification Report :",classification_report(y_test,y_pred3))
print()
print("Confusion Matrix :",confusion_matrix(y_test,y_pred3))
 

Model Accuracy %: 81.22977346278317

Classification Report :               precision    recall  f1-score   support

           0       0.99      0.80      0.88      1105
           1       0.35      0.92      0.51       131

    accuracy                           0.81      1236
   macro avg       0.67      0.86      0.70      1236
weighted avg       0.92      0.81      0.84      1236


Confusion Matrix : [[883 222]
 [ 10 121]]


In [22]:
# model4 = FishersLinearDiscriminant()
# model4.fit(X_train, y_train)

# y_pred4 = model4.classify(X_test)

# IV. Experiments

Apply the classfiers on the data and discuss the results.
Please describe your codes for experiments. You may have subsections of results and discussions here.
Here follows the list that you consider to include:
- the classification results
- plots of classification results 
- model comparision 
- choice of evaluation metrics
- **Must partition data into training and testing**

# Conclusions

Summarize your work here. 
Which classifier do you think the best? 
Discuss the challenges or somethat that you learned. 
If you have any suggestion about the assignment, you can write about it. 

## SoftmaxRegression = 62 %
## LogisticRegression = 89 %
## BayesianLogisticRegression = 81 %
## LeastSquareClassifier = 78 %

### After comparing the accuracies it seems that "Logistic Regression" performs  best . Categorical variables were challanging to to encode.However, I learned how to do that from this assignment.Apologiesfor the late submission. 

# References

No outside references taken